File: acotenoff-jscales/cal.py
#!/usr/bin/env python
# cal implementation
# by Adam Cotenoff (@acotenoff)
#    Jordan Scales (@jdan)
# 18 September 2013

from datetime import date
from sys import argv

# Takes an array of dates (maybe with spaces)
# and displays a grid
# Each item is either 1 or two characters wide
def makeGrid(arr):
    index = 0
    string = ''
    for item in arr:
        # convert numbers to strings
        item = str(item)
        index += 1

        # "pad" our single-character strings
        if len(str(item)) == 1:
            item = ' ' + item
        string += item + ' '

        # add a new line at the end of the week
        if index % 7 == 0:
            string += '\n'

    return string

def main():
    monthDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    monthNames = ['January', 'February', 'March', 'April',
                  'May', 'June', 'July', 'August', 'September',
                  'October', 'November', 'December']
    dayNames = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

    if len(argv) == 3:
        month = int(argv[1])
        year = int(argv[2])
    else:
        today = date.today()
        month = today.month
        year  = today.year

    # first day of the month
    firstOfMonth = date(year, month, 1)
    dayOfWeek = firstOfMonth.weekday()

    # prepend the numbers array
    spacing = [' '] * ((dayOfWeek + 1) % 7)
    numbers = list(range(1, monthDays[month - 1] + 1))

    # deal with leap years
    if month == 2:
        if year % 400 == 0:
            numbers.append(29)
        elif year % 100 == 0:
            pass
        elif year % 4 == 0:
            numbers.append(29)

    # try and center the header (September 2013) as best as we can
    header = monthNames[month - 1] + ' ' + str(year)
    print ' ' * ((20 - len(header)) / 2) + header

    # print the short day names "Su Mo Tu We Th ..."
    print ' '.join(dayNames)

    # print the dates in a grid
    print makeGrid(spacing + numbers)

if __name__ == "__main__":
    main()

File: tmiranda/cal.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


char *monthrel[] = {
	"January",
	"February",
	"March",
	"April",
	"May",
	"June",
	"July",
	"August",
	"September",
	"October",
	"November",
	"December",
};

int monthdays[] = {31, 0, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};

/* Given a day, month, and year what is the day of the week? */
int day_of_week(int day, int month, int year)
{
	int a = (14 - month) / 12;
	int y = year - a;
	int m = month + 12 * a - 2;

	return ((day + y + y / 4 - y / 100 + y / 400 + (31 * m) / 12) % 7);
}

/* Is it a leap year? */
int leap_year(int year)
{
	if (year % 4 == 0)
	{
		if (year % 100 == 0)
		{
			if (year % 400 == 0)
				return (1);
			else
				return (0);
		}
		else
			return (1);
	}
	else
		return (0);
}

int main(int argc, char *argv[])
{
	time_t curtime;
	struct tm *curdate;
	int i, j, fday, tdays, month, year;

	month = atoi(argv[1]);
	year = atoi(argv[2]);

	/* Don't want to enter arguments? Well, fuck you too, buddy. */
	if (month == '\0' || year == 0)
	{
		curtime = time(NULL);
		curdate = localtime(&curtime);
		month = curdate->tm_mon + 1;
		year = curdate->tm_year + 1900;
	}

	puts(""); /* Newline */
	printf("%s %d\n", monthrel[month - 1], year);
	puts("Su Mo Tu We Th Fr Sa");
	puts("--------------------");

	/* Deal with the first week */
	fday = day_of_week(1, month, year);
	for (i = 0, j = 1; i < 8; i++)
	{
		if (i > 0 && day_of_week(j, month, year) == 0)
			break;
		if (i < fday)
			printf("   ");
		else
		{
			printf("%2.d ", j);
			j++;
		}
	}
	puts(""); /* Newline */

	/* Is it a leap year? */
	if (month == 2)
		tdays = leap_year(year) ? 29 : 28;
	else
		tdays = monthdays[month - 1];

	/* Deal with the rest of the month */
	for (i = j; i <= tdays; i++)
	{
		printf("%2.d ", i);
		if (day_of_week(i, month, year) == 6)
			puts("");
	}
	puts(""); /* Newline */

	return (0);
}
